graphql/database/base/Model.js

```diff
@@ -0,0 +1,469 @@
+import moment from 'moment'
+import { get, has } from 'lodash/object'
+import { isObject, isFunction, isNil } from 'lodash/lang'
+
+import {
+  DatabaseConditionalCheckFailedException,
+  DatabaseItemDoesNotExistException,
+  NotImplementedException,
+  UnauthorizedQueryException,
+} from '../../utils/exceptions'
+import { isValidPermissionsOverride } from '../../utils/permissions-overrides'
+import logger from '../../utils/logger'
+
+class Model {
+  constructor(obj) {
+    if (!obj || typeof obj !== 'object') {
+      return
+    }
+    const fieldNames = [].concat(Object.keys(this.constructor.schema), [
+      'created',
+      'updated',
+    ])
+    const customDeserializers = this.constructor.fieldDeserializers
+    const { fieldDefaults } = this.constructor
+
+    // Set properties for each field on the model.
+    // * If `obj[fieldName]` exists, use the value of `obj[fieldName]`.
+    // * Else, if `obj[fieldName]` does not exist, use the default value of
+    //   the field if one exists.
+    // * If both the value and default value are nil, do not set
+    //   the property.
+    fieldNames.forEach(fieldName => {
+      let val = null
+      if (has(obj, fieldName)) {
+        val = obj[fieldName]
+      } else if (has(fieldDefaults, fieldName)) {
+        const fieldDefault = fieldDefaults[fieldName]
+        if (isFunction(fieldDefault)) {
+          val = fieldDefault()
+        } else {
+          val = fieldDefault
+        }
+      }
+      if (!isNil(val)) {
+        // TODO: clean up the logic in this constructor to use immutable variables.
+        // eslint-disable-next-line no-param-reassign
+        obj[fieldName] = val
+      }
+    })
+
+    // Call any custom deserializers on the fields.
+    // * If a custom deserializer exists for that field, call it.
+    // * If the returned value is null or undefined, do not set
+    //   the property.
+    const self = this
+    fieldNames.forEach(fieldName => {
+      let val = null
+      if (has(obj, fieldName)) {
+        val = obj[fieldName]
+      }
+      if (isFunction(get(customDeserializers, fieldName, false))) {
+        const deserializeFunc = customDeserializers[fieldName]
+        val = deserializeFunc(val, obj)
+      }
+      if (!isNil(val)) {
+        self[fieldName] = val
+      }
+    })
+  }
+
+  static register() {
+    // Kind of dumb logic for now, just makes sure all expected internal methods are defined.
+    /* eslint-disable no-unused-expressions */
+    this.schema
```

Comment:
Do I need this? This might be a little dumb.

Reply:
This seems just fine to me. It's like build-time validation for whatever model extends this.

Reply:
If the expectation is to error to the developer, though, maybe check each of these and throw a more useful error message.